js/iprange_test.go

```diff
@@ -0,0 +1,201 @@
+package js
+
+import (
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"math/rand"
+	"net"
+	"testing"
+	"time"
+)
+
+func ipInRange(ip net.IP, ipStart string, ipEnd string) bool {
+	start, end := net.ParseIP(ipStart), net.ParseIP(ipEnd)
+	if start == nil || end == nil {
+		return false
+	}
+	if c := ip.To4(); c != nil {
+		a, b := start.To4(), end.To4()
+		return bytes.Compare(c, a) >= 0 && bytes.Compare(c, b) <= 0
+	}
+	if d := ip.To16(); d != nil {
+		a, b := start.To16(), end.To16()
+		return bytes.Compare(d[:8], a[:8]) >= 0 &&
+			bytes.Compare(d[:8], b[:8]) <= 0 &&
+			bytes.Compare(d[8:], a[8:]) >= 0 &&
+			bytes.Compare(d[8:], b[8:]) <= 0
+	}
+	return false
+}
+
+func TestIpBlockFromRange(t *testing.T) {
+	testdata := map[string]struct {
+		hostCount uint64
+		netCount  uint64
+		ipBlock   string
+	}{
+		"ipv4 range 1": {100, 1, "192.168.0.101-192.168.0.200"},
+		"ipv4 range 2": {101, 1, "192.168.0.100-192.168.0.200"},
+		"ipv6 range 1": {1024, 256, "fd00:1:1:0::0-fd00:1:1:ff::3ff"},
+		"ipv6 range 2": {1023, 254, "fd00:1:1:2::1-fd00:1:1:ff::3ff"},
+	}
+	for name, data := range testdata {
+		data := data
+		t.Run(name, func(t *testing.T) {
+			b := ipBlockFromRange(data.ipBlock)
+			assert.NotNil(t, b)
+			assert.Equal(t, data.hostCount, b.hostN)
+			assert.Equal(t, data.netCount, b.netN)
+		})
+	}
+}
+
+func TestIpBlockFromCIDR(t *testing.T) {
+	testdata := map[string]struct {
+		hostCount uint64
+		netCount  uint64
+		ipBlock   string
+	}{
+		"ipv4 cidr 1": {65536, 1, "192.168.0.0/16"},
+		"ipv4 cidr 2": {65526, 1, "192.168.0.10/16"},
+		"ipv6 cidr 1": {256, 1, "fd00::0/120"},
+		"ipv6 cidr 2": {65536, 1, "fd00::0/112"},
+		"ipv6 cidr 3": {253, 1, "fd00::3/120"},
+		"ipv6 cidr 4": {65534, 1, "fd00::2/112"},
+	}
+	for name, data := range testdata {
+		data := data
+		t.Run(name, func(t *testing.T) {
+			b := ipBlockFromCIDR(data.ipBlock)
+			assert.NotNil(t, b)
+			assert.Equal(t, data.hostCount, b.hostN)
+			assert.Equal(t, data.netCount, b.netN)
+		})
+	}
+}
+
+func TestGetIPBlock(t *testing.T) {
+	testdata := map[string]struct {
+		hostCount uint64
+		netCount  uint64
+		ipBlock   string
+	}{
+		"ipv4 range 1": {100, 1, "192.168.0.101-192.168.0.200"},
+		"ipv4 range 2": {101, 1, "192.168.0.100-192.168.0.200"},
+		"ipv4 cidr 1":  {256, 1, "192.168.0.0/24"},
```

Comment:
But ... there are 254 actuallu usable addresses ... and the code doesn't return `.0` and `.255` unless they are the only IP ... which I think is the correct behaviour. But than this 256 is wrong

Reply:
yes. this just tests for board checking. in practice a user should use  "ipv4 cidr 1":  {254, 1, "192.168.0.01/24|0|254"}